### ISSUE TITLE
Big number issues

### ISSUE THREAD

**Issue by hikarine3** [NectarJS/nectarjs#81]
### Case 1: Number corrupts from certain amount
```
let num = 1;
for(let i=1;i<=100;i++){
	num *=i;
	console.log(num);
}
```
node's output

```
node e.js|head -n 40
1
2
6
24
120
720
5040
40320
362880
3628800
39916800
479001600
6227020800
87178291200
1307674368000
20922789888000
355687428096000
6402373705728000
121645100408832000
2432902008176640000
51090942171709440000
1.1240007277776077e+21
2.585201673888498e+22
6.204484017332394e+23
1.5511210043330986e+25
4.0329146112660565e+26
1.0888869450418352e+28
3.0488834461171384e+29
8.841761993739701e+30
2.6525285981219103e+32
8.222838654177922e+33
2.631308369336935e+35
8.683317618811886e+36
2.9523279903960412e+38
1.0333147966386144e+40
3.719933267899012e+41
1.3763753091226343e+43
5.23022617466601e+44
2.0397882081197442e+46
8.159152832478977e+47
```

nectarjs.js output
```
./e |head -n 40
1
2
6
24
120
720
5040
40320
362880
3628800
39916800
479001600
1932053504
1278945280
2004310016
2004189184
-288522240
-898433024
109641728
-2102132736
-1195114496
-522715136
862453760
-775946240
2076180480
-1853882368
1484783616
-1375731712
-1241513984
1409286144
738197504
-2147483648
-2147483648
0
0
0
0
0
0
0
```

### Case 2: Infinity data type
```
const x = Infinity;
console.log(x);
```
node&deno can execute it but nectar failed for compilation.

```
nectar infinity.js 
[*] Generating source file
[*] Compiling
/Users/hajimekurita/tmp/.nectar/gnan4/infinity.cpp:195:7: error: C++ requires a type specifier for all declarations
const x = Infinity;
~~~~~ ^
/Users/hajimekurita/tmp/.nectar/gnan4/infinity.cpp:195:11: error: use of undeclared identifier 'Infinity'
const x = Infinity;
          ^
2 errors generated.
[!] Compilation error
Error: Command failed: g++ -std=c++11 /Users/hajimekurita/tmp/.nectar/gnan4/infinity.cpp -O1 -fpermissive -w -s   -o /Users/hajimekurita/tmp/infinity
/Users/hajimekurita/tmp/.nectar/gnan4/infinity.cpp:195:7: error: C++ requires a type specifier for all declarations
const x = Infinity;
~~~~~ ^
/Users/hajimekurita/tmp/.nectar/gnan4/infinity.cpp:195:11: error: use of undeclared identifier 'Infinity'
const x = Infinity;
          ^
2 errors generated.

    at checkExecSyncError (child_process.js:616:11)
    at execSync (child_process.js:652:15)
    at Compiler.Compile (/Users/hajimekurita/.nodebrew/node/v12.4.0/lib/node_modules/nectarjs/compiler/native/compiler.js:394:3)
    at /Users/hajimekurita/.nodebrew/node/v12.4.0/lib/node_modules/nectarjs/nectar.js:336:13
    at FSReqCallback.readFileAfterClose [as oncomplete] (internal/fs/read_file_context.js:61:3) {
  status: 1,
  signal: null,
  output: [
    null,
    <Buffer >,
    <Buffer 2f 55 73 65 72 73 2f 68 61 6a 69 6d 65 6b 75 72 69 74 61 2f 74 6d 70 2f 2e 6e 65 63 74 61 72 2f 67 6e 61 6e 34 2f 69 6e 66 69 6e 69 74 79 2e 63 70 70 ... 252 more bytes>
  ],
  pid: 820,
  stdout: <Buffer >,
  stderr: <Buffer 2f 55 73 65 72 73 2f 68 61 6a 69 6d 65 6b 75 72 69 74 61 2f 74 6d 70 2f 2e 6e 65 63 74 61 72 2f 67 6e 61 6e 34 2f 69 6e 66 69 6e 69 74 79 2e 63 70 70 ... 252 more bytes>
}
```

**Comment by adrien-thierry** [comment]
fb9d8a8 shoud solve the problem
to address big numbers efficiently, it's possible to force double with `var i = 1.0`

Infinity now exists in NectarJS

**Comment by hikarine3** [comment]
I couldn't confirm fixing in version 0.6.43 which I install through npm.
Maybe it is in not released version, isn't it?

nectar --version;nectar --preset speed e.js; ./e |head -n 40
NectarJS v0.6.43
[*] Generating source file
[*] Compiling
1
2
6
24
120
720
5040
40320
362880
3628800
39916800
479001600
1932053504
1278945280
2004310016
2004189184
-288522240
-898433024
109641728
-2102132736
-1195114496
-522715136
862453760
-775946240
2076180480
-1853882368
1484783616
-1375731712
-1241513984
1409286144
738197504
-2147483648
-2147483648
0
0
0
0
0
0
0

**Comment by adrien-thierry** [comment]
Could you try with :

`let num = 1.0`

**Comment by hikarine3** [comment]
Here is the result.
Compared to node & deno, I feel it is too early starting e+ expression in the case of nectarjs.
==
nectar --preset speed e.js; ./e |head -n 40
[*] Generating source file
[*] Compiling
1
2
6
24
120
720
5040
40320
362880
3.6288e+06
3.99168e+07
4.79002e+08
6.22702e+09
8.71783e+10
1.30767e+12
2.09228e+13
3.55687e+14
6.40237e+15
1.21645e+17
2.4329e+18
5.10909e+19
1.124e+21
2.5852e+22
6.20448e+23
1.55112e+25
4.03291e+26
1.08889e+28
3.04888e+29
8.84176e+30
2.65253e+32
8.22284e+33
2.63131e+35
8.68332e+36
2.95233e+38
1.03331e+40
3.71993e+41
1.37638e+43
5.23023e+44
2.03979e+46
8.15915e+47
HajimenoMacBook-ea:tmp hajimekurita$ node e.js |head -n 40
1
2
6
24
120
720
5040
40320
362880
3628800
39916800
479001600
6227020800
87178291200
1307674368000
20922789888000
355687428096000
6402373705728000
121645100408832000
2432902008176640000
51090942171709440000
1.1240007277776077e+21
2.585201673888498e+22
6.204484017332394e+23
1.5511210043330986e+25
4.0329146112660565e+26
1.0888869450418352e+28
3.0488834461171384e+29
8.841761993739701e+30
2.6525285981219103e+32
8.222838654177922e+33
2.631308369336935e+35
8.683317618811886e+36
2.9523279903960412e+38
1.0333147966386144e+40
3.719933267899012e+41
1.3763753091226343e+43
5.23022617466601e+44
2.0397882081197442e+46
8.159152832478977e+47
HajimenoMacBook-ea:tmp hajimekurita$ deno e.js |head -n 40
1
2
6
24
120
720
5040
40320
362880
3628800
39916800
479001600
6227020800
87178291200
1307674368000
20922789888000
355687428096000
6402373705728000
121645100408832000
2432902008176640000
51090942171709440000
1.1240007277776077e+21
2.585201673888498e+22
6.204484017332394e+23
1.5511210043330986e+25
4.0329146112660565e+26
1.0888869450418352e+28
3.0488834461171384e+29
8.841761993739701e+30
2.6525285981219103e+32
8.222838654177922e+33
2.631308369336935e+35
8.683317618811886e+36
2.9523279903960412e+38
1.0333147966386144e+40
3.719933267899012e+41
1.3763753091226343e+43
5.23022617466601e+44
2.0397882081197442e+46
8.159152832478977e+47

**Comment by adrien-thierry** [comment]
For better precision I will need to create a custom number class to properly manage integer / float / double / big number without overflow and with good perf

**Comment by Saiv46** [comment]
@hikarine3
NectarJS is being refactored. I wrote a custom number class so numbers won't overflow (until `MAX_SAFE_INTEGER` of course).

Thank you.

**Comment by hikarine3** [comment]
Sorry but I think still the result is same if I use nectarjs which can be gotten through npm.
====
# In the case let num = 0;
 nectar --version;nectar --preset speed e.js; ./e |head -n 40
NectarJS v0.7.10
[*] Generating source file
[*] Compiling with preset: speed
1
2
6
24
120
720
5040
40320
362880
3628800
39916800
479001600
1932053504
1278945280
2004310016
2004189184
-288522240

====
# In the case let num = 1.0
nectar --version;nectar --preset speed e.js; ./e |head -n 40
NectarJS v0.7.10
[*] Generating source file
[*] Compiling with preset: speed
1
2
6
24
120
720
5040
40320
362880
3.6288e+06
3.99168e+07
4.79002e+08
6.22702e+09
8.71783e+10
1.30767e+12
2.09228e+13

**Comment by adrien-thierry** [comment]
The develop branch is not merget yet with the master/npm one, we will do it soon

**Comment by adrien-thierry** [comment]
Big number precision solved in develop branch

**Comment by hikarine3** [comment]
Number looks O.K for var i=1 and var i=1.0 but
I can see "." at the tail of number from the middle of the script, which I think it shouldn't appear.
```
1
2
6
24
120
720
5040
40320
362880
3628800
39916800
479001600
6227020800.
```

hajimekurita@HajimenoMacBook-Air nectarjs % git branch;
* develop
  master
hajimekurita@HajimenoMacBook-Air nectarjs % git branch;node nectar.js e.js

hajimekurita@HajimenoMacBook-Air nectarjs % git branch;node nectar.js e.js;./e;
* develop
  master
[*] Generating source file
[*] Compiling with preset: none
ld: warning: option -s is obsolete and being ignored
1
2
6
24
120
720
5040
40320
362880
3628800
39916800
479001600
6227020800.
87178291200.
1307674368000.
20922789888000.
355687428096000.
6402373705728000.
121645100408832000.
2432902008176640000.
51090942171709440000.
1124000727777607680000.
25852016738884978212864.
620448401733239409999872.
15511210043330986055303168.
403291461126605650322784256.
10888869450418351940239884288.
304888344611713836734530715648.
8841761993739700772720181510144.
265252859812191032188804700045312.
8222838654177922430198509928972288.
263130836933693517766352317727113216.
8683317618811885938715673895318323200.
295232799039604119555149671006000381952.
10333147966386144222209170348167175077888.
371993326789901177492420297158468206329856.
13763753091226343102992036262845720547033088.
523022617466601037913697377988137380787257344.
20397882081197441587828472941238084160318341120.
815915283247897683795548521301193790359984930816.
33452526613163802763987613764361857922667238129664.
1405006117752879788779635797590784832178972610527232.
60415263063373834074440829285578945930237590418489344.
2658271574788448529134213028096241889243150262529425408.
119622220865480188574992723157469373503186265858579103744.
5502622159812088456668950435842974564586819473162983440384.
258623241511168177673491006652997026552325199826237836492800.
12413915592536072528327568319343857274511609591659416151654400.
608281864034267522488601608116731623168777542102418391010639872.
30414093201713375576366966406747986832057064836514787179557289984.
1551118753287382189470754582685817365323346291853046617899802820608.
80658175170943876845634591553351679477960544579306048386139594686464.
4274883284060025484791254765342395718256495012315011061486797910441984.
230843697339241379243718839060267085502544784965628964557765331531071488.
12696403353658276446882823840816011312245221598828319560272916152712167424.
710998587804863481025438135085696633485732409534385895375283304551881375744.
40526919504877220527556156789809444757511993541235911846782577699372834750464.
2350561331282878906297796280456247634956966273955390268712005058924708557225984.
138683118545689864933221185143853352853533809868133429504739525869642019130834944.
8320987112741391580056396102959641077457945541076708813599085350531187384917164032.
507580213877224835833540161373088490724281389843871724559898414118829028410677788672.
31469973260387939390320343330721249710233204778005956144519390914718240063804258910208.
1982608315404440084965732774767545707658109829136018902789196017241837351744329178152960.
126886932185884165437806897585122925290119029064705209778508545103477590511637067401789440.
8247650592082471516735380327295020523842257210146637473076098881993433291790288339528056832.
544344939077443069445496060275635856761283034568718387417404234993819829995466026946857533440.
36471110918188683221214362054827498508015278133658067038296405766134083781086959639263732301824.
2480035542436830547970901153987107983847555399761061789915503815309070879417337773547217359994880.
171122452428141297375735434272073448876652721480628511030304905066123383956194496253690059725733888.
11978571669969890269925854460558840225267029209529303278944419871214396524861374498691473966836482048.
850478588567862176139364498862283450363106657876759119884137546969850291984346623845721803156845756416.
61234458376886076682034243918084408426143679367126656631657903381829221022872956916891969827292894461952.
4470115461512683367030518111879159855011125453675376127499372976904192242294440576507107183576895048384512.
330788544151938558975078458606627397928594841525087028177611631961228972749086355889619432768006702663467008.
24809140811395391401649674453868616759922516881580717884144963618612492793329521656999844718186305916810297344.
1885494701666049846649767567286674986020753759697889931196791720482648043560619012598537844549685032003930423296.
145183092028285837925033723190960213624220326225543510376814897182582303110787462992309948307745678711432381726720.
11324281178206294606285193764734547659641544873910049469239570110699644621282776159978832493218689331409071173009408.
894618213078297291394536105678124660091699861979864830895281485890971416487504167917951760559283842969422038852173824.
71569457046263778832073404098641551692451427821500630228331524401978643519022131505852398484420816675798776564959674368.
5797126020747365547859207609316955153302418317114924299299934140244381749043408420663995193999459259329102516576025837568.
475364333701283981804950871934204857403260987909684614932289567004882674634326008655216234173410083475042065689611178344448.
39455239697206569095363763848575524105091557652834963035068520647090338762195044371133838774594273621857211161281310377377792.
3314240134565351991893962785187002255986138585985099085000359647021178112607661449751964466234594461331925608329126314254532608.
281710411438054936145340070063731769270618697594701234453923080571636972801512907171791271651453851961008097129997800044545179648.
24227095383672724277628115968482030522825707406365319023898287641277303090920578351271773357158562842391230015963508220979549569024.
2107757298379526908723379823722428723253356281476796516744835439819327865545820130415624631892617112325588092315921658107822337949696.
185482642257398355359015441641340379717002520724001675848028952155157918424578282117944000727442910232086550560628100855367408841392128.
16507955160908452497218052643056785820348586593118454131228292379023767876373998621963681432374396360185860939835599722920061648806871040.
1485715964481760688598126444658390468648504385338494631194665214503009929244428884407664620677918947601692538663274334960161140774153486336.
135200152767840228116141248983464474000315190336164090023053168975093962659595589494945258181259227289565007146009064474002163016890673266688.
12438414054641300055918190849808704283732243800785472203727097932000687811897290487171609691949473988773572358313712866199772466321053448142848.
1156772507081640872708168771053910362702557453046286765300873130615713476227390429979366250504125810956242157790289615727775585913909863306493952.
108736615665674240994816729183762001752135081532694567149618171011945851076444857687577437120278217589896360543234882091889200965381187714945122304.
10329978488239052206885505130495304991006115078121225426977331552932345209983325487893772190631276157126501537780698430576453773283127019334195478528.
991677934870949102715849029478410597802063033500204435194993914996262675344674144417713144475342926858025486847392744430654763018452435547907969515520.
96192759682482062236598631563798937437476306361515295860273049067319419226943192827878886900710340579037421510433530649010990406523708314612471414390784.
9426890448883242029410148360874034376137592466180063696357188182614769297217373775790759257383412737431326439501183709769874985637770333212700442263289856.
933262154439440960911604687726529403237621654151826305939361630078862160424520003803285166480957861005701317510617187267217623578139262988057343784065695744.
93326215443944102188325606108575267240944254854960571509166910400407995064242937148632694030450512898042989296944474898258737204311236641477561877016501813248.

**Comment by adrien-thierry** [comment]
I updated it, almost same behavior as V8 now: 

1
2
6
24
120
720
5040
40320
362880
3628800
39916800
479001600
6227020800
87178291200
1307674368000
20922789888000
355687428096000
6402373705728000
121645100408832000
2432902008176640000
51090942171709440000
1.124000727777607e+21
2.585201673888497e+22
6.204484017332394e+23
1.551121004333098e+25
4.032914611266056e+26
1.088886945041835e+28
...

We just have to print correctly infinite number and it's good

**Comment by adrien-thierry** [comment]
Corrected in v0.7.23

**Comment by krazykoder** [comment]
> Corrected in v0.7.23

Sorry I am a noob here. How can i download / use this version v0.7.23 with bignumber fix. 
npm is still on 0.7.115
Appreciate if you can update npm. Thank you.
